File: AdminApi.Sdk/Endpoints/Tiers/TiersEndpoint.cs
using Backbone.AdminApi.Sdk.Endpoints.Tiers.Types;
using Backbone.AdminApi.Sdk.Endpoints.Tiers.Types.Requests;
using Backbone.AdminApi.Sdk.Endpoints.Tiers.Types.Responses;
using Backbone.BuildingBlocks.SDK.Endpoints.Common;
using Backbone.BuildingBlocks.SDK.Endpoints.Common.Types;

namespace Backbone.AdminApi.Sdk.Endpoints.Tiers;

public class TiersEndpoint(EndpointClient client) : Endpoint(client)
{
    public async Task<ApiResponse<ListTiersResponse>> ListTiers() => await _client.Get<ListTiersResponse>("Tiers");

    public async Task<ApiResponse<TierDetails>> GetTier(string id) => await _client.Get<TierDetails>($"Tiers/{id}");

    public async Task<ApiResponse<Tier>> CreateTier(CreateTierRequest request) => await _client.Post<Tier>("Tiers", request);

    public async Task<ApiResponse<EmptyResponse>> DeleteTier(string id) => await _client.Delete<EmptyResponse>($"Tiers/{id}");

    public async Task<ApiResponse<TierQuotaDefinition>> AddTierQuota(string id, CreateQuotaForTierRequest request)
        => await _client.Post<TierQuotaDefinition>($"Tiers/{id}/Quotas", request);

    public async Task<ApiResponse<EmptyResponse>> DeleteTierQuota(string id, string quotaId)
        => await _client.Delete<EmptyResponse>($"Tiers/{id}/Quotas/{quotaId}");
}

File: Applications/AdminUi/packages/admin_api_sdk/lib/src/endpoints/tiers_endpoint.dart
import 'package:admin_api_types/admin_api_types.dart';

import '../types/types.dart';
import 'endpoint.dart';

class TiersEndpoint extends Endpoint {
  TiersEndpoint(super.dio);

  Future<ApiResponse<List<TierOverview>>> getTiers() => get('/api/v1/Tiers', transformer: (e) => (e as List).map(TierOverview.fromJson).toList());

  Future<ApiResponse<Tier>> createTier({required String name}) => post('/api/v1/Tiers', data: {'name': name}, transformer: Tier.fromJson);

  Future<ApiResponse<TierDetails>> getTier(String tierId) => get('/api/v1/Tiers/$tierId', transformer: TierDetails.fromJson);

  Future<ApiResponse<void>> deleteTier(String tierId) =>
      delete('/api/v1/Tiers/$tierId', expectedStatus: 204, transformer: (e) {}, allowEmptyResponse: true);
}
